### PR TITLE
Implement actor checkpointing

### PR DESCRIPTION
**Pull request by raulchen**
## What do these changes do?

This PR implements actor checkpointing. To enable checkpointing, users should inherent their actor classes from the `Checkpointable` interface. See `Checkpointable` definition in `actor.py/Checkpointable.java` for more details. 

## Related issue number

#3818

### CODE REVIEW

**Comment by raulchen** [comment]
@stephanie-wang @ujvl This PR is ready for review.

**Comment by ujvl** [comment]
Thanks! I'll take a pass through this later today.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Thanks, this looks great so far!

**Comment by stephanie-wang** [comment]
@ericl, can you take a look at the Python API?

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -1693,35 +1780,73 @@ void NodeManager::FinishAssignedActorTask(Worker &worker, const Task &task) {
...
+          for (const auto &entry : actor_entry->second.GetDummyObjects()) {
+            HandleObjectLocal(entry.first);
+          }
+        },
+        [actor_id](ray::gcs::AsyncGcsClient *client, const UniqueID &checkpoint_id) {
+          RAY_LOG(WARNING) << "Couldn't find checkpoint " << checkpoint_id
+                           << " for actor " << actor_id << " in GCS. This is likely"
+                           << " because the worker sent us a wrong or expired"
+                           << " checkpoint id.";
+          // TODO(hchen): what should we do here? Notify or kill the actor?
```
[the remark this answers is not in the document]
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: I just realized that I've already checked whether the returned checkpoint id is valid at the front end. I can simply do a `RAY_LOG(FATAL)` here.

File: python/ray/includes/unique_ids.pxi [hunk 2]
```diff
@@ -234,6 +246,29 @@ cdef class ActorHandleID(UniqueID):
         return "ActorHandleID(" + self.hex() + ")"
 
 
+cdef class ActorCheckpointID(UniqueID):
+
+    def __init__(self, id):
+        if not id:
+            self.data = CUniqueID()
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/includes/unique_ids.pxi]: Ahh I see, I didn't realize what the rest of this file looked like :)

Sure, let's stick with what you have here for now.

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -519,36 +553,35 @@ void NodeManager::HandleActorStateTransition(const ActorID &actor_id,
...
-      // we can remove the task from the queue and forget its dependency on the
-      // actor creation task.
-      RAY_CHECK(task_dependency_manager_.UnsubscribeDependencies(
-          method.GetTaskSpecification().TaskId()));
-      // The task's uncommitted lineage was already added to the local lineage
-      // cache upon the initial submission, so it's okay to resubmit it with an
-      // empty lineage this time.
-      SubmitTask(method, Lineage());
+    // The actor's location is now known.
+    bool resumed_from_checkpoint = checkpoint_id_to_restore_.count(actor_id) > 0;
```
Comment (ujvl) [inline src/ray/raylet/node_manager.cc]: nit, slight preference for `find` over `count` for clarity with `unordered_map`

**Review by ujvl — COMMENTED** [review]
Couple minor comments, otherwise looks good!

File: src/ray/gcs/tables.cc [hunk 4]
```diff
@@ -438,6 +440,40 @@ std::string ClientTable::DebugString() const {
...
+        std::make_shared<ActorCheckpointIdDataT>(data);
+    copy->timestamps.push_back(current_sys_time_ms());
+    copy->checkpoint_ids.push_back(checkpoint_id.binary());
+    auto num_to_keep = RayConfig::instance().num_actor_checkpoints_to_keep();
+    while (copy->timestamps.size() > num_to_keep) {
+      // Delete the checkpoint from actor checkpoint table.
+      const auto &checkpoint_id = UniqueID::from_binary(*copy->checkpoint_ids.begin());
+      RAY_LOG(DEBUG) << "Deleting checkpoint " << checkpoint_id << " for actor " << actor_id;
+      copy->timestamps.erase(copy->timestamps.begin());
+      copy->checkpoint_ids.erase(copy->checkpoint_ids.begin());
```
Comment (jovany-wang) [inline src/ray/gcs/tables.cc]: Since checkpoint_ids is a long string concatenated objects:
```
checkpoint_ids: [string];
```
we should not erase the `begin()`.
`begin()` is just the first char of the string.

File: src/ray/gcs/tables.cc [hunk 5]
```diff
@@ -438,6 +440,40 @@ std::string ClientTable::DebugString() const {
...
+                             ray::gcs::AsyncGcsClient *client, const UniqueID &id,
+                             const ActorCheckpointIdDataT &data) {
+    std::shared_ptr<ActorCheckpointIdDataT> copy =
+        std::make_shared<ActorCheckpointIdDataT>(data);
+    copy->timestamps.push_back(current_sys_time_ms());
+    copy->checkpoint_ids.push_back(checkpoint_id.binary());
+    auto num_to_keep = RayConfig::instance().num_actor_checkpoints_to_keep();
+    while (copy->timestamps.size() > num_to_keep) {
+      // Delete the checkpoint from actor checkpoint table.
+      const auto &checkpoint_id = UniqueID::from_binary(*copy->checkpoint_ids.begin());
```
Comment (jovany-wang) [inline src/ray/gcs/tables.cc]: Also here.

File: src/ray/gcs/format/gcs.fbs [hunk 6]
```diff
@@ -310,3 +312,33 @@ table DriverTableData {
...
+  // The numbers of dependencies for the above unreleased dummy objects.
+  num_dummy_object_dependencies: [int];
+}
+
+// This table stores the actor-to-available-checkpoint-ids mapping.
+table ActorCheckpointIdData {
+  // ID of this actor.
+  actor_id: string;
+  // A list of the available checkpoint IDs for this actor.
+  checkpoint_ids: [string];
```
Comment (jovany-wang) [inline src/ray/gcs/format/gcs.fbs]: Since we often insert `id` to `checkpoint_ids `, and remove elements from `checkpoint_ids`.
Is it better to define a Objects table like this?
```
table Objects {
    Object[] objects;
}
table Object {
    byte[] object;
}
```

File: src/ray/raylet/node_manager.cc [hunk 7]
```diff
@@ -519,36 +553,35 @@ void NodeManager::HandleActorStateTransition(const ActorID &actor_id,
...
+    if (!resumed_from_checkpoint) {
+      // If the actor wasn't resumed from a checkpoint, just resubmit the waiting tasks.
+      ResubmitTasksWaitingForActorCreation(actor_id);
+    } else {
+      // If this actor was resumed from a checkpoint, look up the checkpoint in GCS,
+      // retore actor state, and resubmit the waiting tasks.
+      const auto checkpoint_id = checkpoint_id_to_restore_[actor_id];
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Hmm I'm not sure if this will work in a distributed setting, since the publish could still go out to other nodes, which will resubmit their cached tasks, and that could potentially happen before we restore the checkpoint here.

**Comment by stephanie-wang** [comment]
One question about the API: the backend only keeps around the last n checkpoints for an actor, but it seems like we should notify the application when older checkpoints get garbage-collected in the GCS, right? Have you thought about how we should do that?

File: src/ray/gcs/tables.cc (hunk 4 above)
Comment (raulchen) [reply src/ray/gcs/tables.cc]: Currently, `checkpoint_ids` is defined as a list of strings, using `begin()` is fine. But I should concatenate them as one single string. Thanks for reminding

File: src/ray/gcs/format/gcs.fbs [hunk 8]
```diff
@@ -310,3 +312,33 @@ table DriverTableData {
...
+  // The frontier dependencies of the above handles.
+  frontier_dependencies: [string];
+  // A list of unreleased dummy objects from this actor.
+  unreleased_dummy_objects: [string];
+  // The numbers of dependencies for the above unreleased dummy objects.
+  num_dummy_object_dependencies: [int];
+}
+
+// This table stores the actor-to-available-checkpoint-ids mapping.
+table ActorCheckpointIdData {
```
[the remark this answers is not in the document]
Comment (raulchen) [reply src/ray/gcs/format/gcs.fbs]: I didn't do that because of efficiency. `checkpoint_ids` might be modified pretty frequently (currently we need to do checkpoint for every task to bypass replaying tasks). 
`ActorTable` is a log, if we put `checkpoint_ids` in `ActorTable`, we'll need to keep all the histories, which isn't necessary.

File: python/ray/worker.py [hunk 9]
```diff
@@ -959,6 +960,56 @@ def _wait_for_and_process_task(self, task):
...
+                        "`load_checkpoint must return a checkpoint id that " +
+                        "exists in the `available_checkpoints` list, or None.")
+                    assert any(checkpoint_id == checkpoint.checkpoint_id
+                               for checkpoint in checkpoints), msg
+                    # Notify raylet that this actor has been resumed from
+                    # a checkpoint.
+                    self.raylet_client.notify_actor_resumed_from_checkpoint(
+                        actor_id, checkpoint_id)
+        elif is_actor_task:
+            self._num_tasks_since_last_checkpoint += 1
```
[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/worker.py]: No. Because one worker can only have at most one actor at the same time.

File: src/ray/raylet/node_manager.cc (hunk 3 above)
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: I use `count` because it's shorter and can fit in one line. Do you know any pros of using `find` over `count`? I think in terms of efficiency, they should be the same.

**Comment by raulchen** [comment]
> One question about the API: the backend only keeps around the last n checkpoints for an actor, but it seems like we should notify the application when older checkpoints get garbage-collected in the GCS, right? Have you thought about how we should do that?

We pass in the `available_checkpoints` parameter to `load_checkpoint`, so users can know which checkpoints are still available. This should be okay?

File: src/ray/raylet/node_manager.cc (hunk 7 above)
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: Ah, yes, that problem could still happen.

**Comment by stephanie-wang** [comment]
> > One question about the API: the backend only keeps around the last n checkpoints for an actor, but it seems like we should notify the application when older checkpoints get garbage-collected in the GCS, right? Have you thought about how we should do that?
> 
> We pass in the `available_checkpoints` parameter to `load_checkpoint`, so users can know which checkpoints are still available. This should be okay?

I meant that we should probably let the user know when checkpoints get GC'ed in the GCS, so that they can GC the application checkpoint data.

**Comment by raulchen** [comment]
@stephanie-wang I see. 
I can make the `num_checkpoitns_to_keep` config available to user code. When they save a new app-level checkpoint, they can clean up the old ones. Is this okay?

File: python/ray/worker.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ujvl) [reply python/ray/worker.py]: Might still be cleaner to move those variables to the actor in case that assumption changes in the future, unless there's a good reason for keeping it here.

**Comment by stephanie-wang** [comment]
> > When they save a new app-level checkpoint, they can clean up the old ones. Is this okay?
>  
> You could add another callback function to the `Checkpointable` interface like `cleanup_checkpoint(checkpoint_id)` that's implemented by the user if they want to do GC, we can call it in `_handle_actor_checkpoint` after `save_checkpoint`. That way they don't need to explicitly assume which checkpoint is garbage collected when they take a new checkpoint.

+1

Actually, we should probably advise the user about how to write checkpoint data. For instance, if they write the checkpoint data in place (like you currently have in the Python test), it could break if `num_actor_checkpoints_to_keep=1` since the application checkpoint is not atomic with the backend checkpoint. Not to say that we should enforce that in this PR, but it'd be good to add this to the online documentation.

File: src/ray/ray_config_def.h [hunk 10]
```diff
@@ -135,3 +135,6 @@ RAY_CONFIG(int, num_workers_per_process, 1);
 
 /// Maximum timeout in milliseconds within which a task lease must be renewed.
 RAY_CONFIG(int64_t, max_task_lease_timeout_ms, 60000);
+
+/// Maximum number of checkpoints to keep in GCS for an actor.
+RAY_CONFIG(uint32_t, num_actor_checkpoints_to_keep, 200);
```
Comment (stephanie-wang) [inline src/ray/ray_config_def.h]: How about defaulting this to 2? If the application is writing its checkpoint data in place, then that would guarantee that the stored checkpoint is always in the backend's `available_checkpoints`. If the application isn't writing its checkpoint data in place and therefore needs to GC old checkpoints, then this would minimize the amount of checkpoint data per actor.

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -1693,35 +1757,88 @@ void NodeManager::FinishAssignedActorTask(Worker &worker, const Task &task) {
...
+          }
+          for (const auto &task : local_queues_.RemoveTasks(waiting_actor_task_ids)) {
+            if (!lineage_cache_.RemoveWaitingTask(
+                    task.GetTaskSpecification().TaskId())) {
+              RAY_LOG(WARNING) << "Task " << task.GetTaskSpecification().TaskId()
+                               << " already removed from the lineage cache. This is most "
+                                  "likely due to reconstruction.";
+            }
+            SubmitTask(task, Lineage());
+          }
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: @stephanie-wang I ended up fixing the issue by resubmitting the waiting task here.

File: python/ray/actor.py [hunk 12]
```diff
@@ -900,3 +904,125 @@ def __ray_checkpoint_restore__(self):
...
+            ID of one checkpoint in this list, or None. Otherwise, an exception
+            will be raised.
+        Returns:
+            The ID of the checkpoint from which the actor was resumed, or None
+            if the actor should restart from the beginning.
+        """
+        pass
+
+    @abstractmethod
+    def checkpoint_expired(self, checkpoint_id):
```
Comment (raulchen) [inline python/ray/actor.py]: @ujvl @stephanie-wang added a `checkpoint_expired` callback here.

File: src/ray/ray_config_def.h [hunk 13]
```diff
@@ -135,3 +135,9 @@ RAY_CONFIG(int, num_workers_per_process, 1);
 
 /// Maximum timeout in milliseconds within which a task lease must be renewed.
 RAY_CONFIG(int64_t, max_task_lease_timeout_ms, 60000);
+
+/// Maximum number of checkpoints to keep in GCS for an actor.
+/// Note: this number should be set to at least 2. Because saving a application
+/// checkpoint isn't atomic with saving the backend checkpoint, and it will break
+/// if this number is set to 1 and users save application checkpoints in place.
```
Comment (raulchen) [inline src/ray/ray_config_def.h]: @stephanie-wang I added  a note here to warn about setting `num_actor_checkpoints_to_keep=1`.

**Comment by raulchen** [comment]
@jovany-wang Java part is done, please help take a look. thanks

File: python/ray/actor.py (hunk 12 above)
Comment (ujvl) [reply python/ray/actor.py]: think you need to pass in the `actor_id` in here as well if it's used to locate the data (since `save_checkpoint` may use it in that way).

File: src/ray/gcs/format/gcs.fbs [hunk 14]
```diff
@@ -337,8 +337,9 @@ table ActorCheckpointData {
 table ActorCheckpointIdData {
   // ID of this actor.
   actor_id: string;
-  // A list of the available checkpoint IDs for this actor.
-  checkpoint_ids: [string];
+  // IDs of this actor's available checkpoints.
+  // Note, this is a long string that concatenates all the IDs.
+  checkpoint_ids: string;
```
Comment (stephanie-wang) [inline src/ray/gcs/format/gcs.fbs]: Why the change from `[string]` to `string`? This seems more confusing...

File: test/actor_test.py [hunk 15]
```diff
@@ -28,6 +28,8 @@ def ray_start_regular(request):
     # Start the Ray processes.
     ray.init(
         num_cpus=1,
+        redirect_output=True,
```
Comment (stephanie-wang) [inline test/actor_test.py]: I don't think these are necessary.

File: src/ray/raylet/node_manager.cc (hunk 11 above)
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc]: Hmm okay, I think I prefer the other solution I mentioned because it seems cleaner, but maybe I'm wrong. I'll give it a shot and push to the PR if it seems doable.

File: src/ray/raylet/actor_registration.cc [hunk 16]
```diff
@@ -77,6 +77,53 @@ void ActorRegistration::AddHandle(const ActorHandleID &handle_id,
 
 int ActorRegistration::NumHandles() const { return frontier_.size(); }
 
+void ActorRegistration::RestoreFrontier(const ActorCheckpointDataT &checkpoint_data) {
+  // Restore `execution_dependency_`.
+  execution_dependency_ = ObjectID::from_binary(checkpoint_data.execution_dependency);
+  // Restore `frontier_`.
```
Comment (stephanie-wang) [inline src/ray/raylet/actor_registration.cc]: We should probably clear `frontier_` and `dummy_objects_` before restoring. Also, we should make sure to release any of the objects in `dummy_objects_` before the restore.

File: python/ray/worker.py [hunk 17]
```diff
@@ -959,6 +960,62 @@ def _wait_for_and_process_task(self, task):
...
+                now - self._last_checkpoint_timestamp)
+            if actor.should_checkpoint(checkpoint_context):
+                # If `should_checkpoint` returns True, notify raylet to prepare
+                # a checkpoint and then call `save_checkpoint`.
+                checkpoint_id = self.raylet_client.prepare_actor_checkpoint(
+                    actor_id)
+                self._actor_checkpoint_ids.append(checkpoint_id)
+                if (len(self._actor_checkpoint_ids) >
+                        ray._config.num_actor_checkpoints_to_keep()):
+                    actor.checkpoint_expired(self._actor_checkpoint_ids[0])
```
Comment (stephanie-wang) [inline python/ray/worker.py]: nit: You can do these two lines as `actor.checkpoint_expired(self._actor_checkpoint_ids.pop(0))`

**Comment by stephanie-wang** [comment]
Okay, I pushed some changes to hopefully address the checkpoint resume issue. Now it waits to publish the actor creation until the checkpoint has resumed. Do you want to take a look, @raulchen?

File: test/actor_test.py (hunk 15 above)
Comment (raulchen) [reply test/actor_test.py]: ah, this was for debugging. I'll remove it.

File: src/ray/gcs/format/gcs.fbs (hunk 14 above)
Comment (raulchen) [reply src/ray/gcs/format/gcs.fbs]: it's because flatbuffer generated Java class cannot get a single id from the list. We've already changed other `[string]` to long strings. See https://github.com/ray-project/ray/pull/3749

File: src/ray/gcs/tables.cc [hunk 18]
```diff
@@ -38,6 +40,8 @@ namespace gcs {
 template <typename ID, typename Data>
 Status Log<ID, Data>::Append(const JobID &job_id, const ID &id,
                              std::shared_ptr<DataT> &dataT, const WriteCallback &done) {
+  RAY_LOG(DEBUG) << client_->client_table().GetLocalClientId() << " Append to table "
+                 << EnumNameTablePrefix(prefix_) << " key " << id;
```
Comment (raulchen) [inline src/ray/gcs/tables.cc]: @stephanie-wang do we still want to keep these logs? I'm not sure if they will be too verbose.

**Review by jovany-wang — COMMENTED** [review]
The Java part looks good to me.

File: java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java [hunk 19]
```diff
@@ -87,6 +104,16 @@ public void start() throws Exception {
...
   }
 
+  private void initRedisClients() {
+    redisClient = new RedisClient(rayConfig.getRedisAddress());
+    int numRedisShards = Integer.valueOf(redisClient.get("NumRedisShards", null));
+    List<String> addresses = redisClient.lrange("RedisShards", 0, -1);
+    Preconditions.checkState(numRedisShards == addresses.size());
+    redisClients = addresses.stream().map(RedisClient::new)
+        .collect(Collectors.toList());
+    redisClients.add(redisClient);
```
[the remark this answers is not in the document]
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java]: thanks, I added a TODO comment in `getCheckpointsForActor`. I think the correct way is to add a `GlobalState` in Java.

**Comment by raulchen** [comment]
Thanks everyone. I think all comments are addressed. Please let me know if I missed something.

**Review by jovany-wang — APPROVED** [review]

File: src/ray/gcs/tables.cc (hunk 18 above)
Comment (stephanie-wang) [reply src/ray/gcs/tables.cc]: Thanks, yeah, I'll delete these!

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/actor.py [hunk 20]
```diff
@@ -832,58 +727,142 @@ def __ray_checkpoint__(self):
...
+    def should_checkpoint(self, checkpoint_context):
+        """Whether this actor needs to be checkpointed.
+
+        This method will be called after every task. You should implement this
+        callback to decide whether this actor needs to be checkpointed at this
+        time, based on the checkpoint context, or any other factors.
+
+        Args:
+            checkpoint_context: A namedtuple that contains info about last
+            checkpoint.
```
Comment (pschafhalter) [inline python/ray/actor.py]: Should be indented.

File: python/ray/actor.py [hunk 21]
```diff
@@ -832,58 +727,142 @@ def __ray_checkpoint__(self):
...
+        should restore actor's state and return the checkpoint ID. Otherwise,
+        it should do nothing and return None.
+
+        Args:
+            actor_id: Actor’s ID.
+            available_checkpoints: A list of `Checkpoint` namedtuples that
+            contains all available checkpoint IDs and their timestamps, sorted
+            by timestamp in descending order. Note, this method must return the
+            ID of one checkpoint in this list, or None. Otherwise, an exception
+            will be raised.
```
Comment (pschafhalter) [inline python/ray/actor.py]: Should be indented as well.

File: python/ray/actor.py [hunk 22]
```diff
@@ -832,58 +727,142 @@ def __ray_checkpoint__(self):
...
+        This method will be called when an checkpoint is expired. You should
+        implement this method to delete your application checkpoint data.
+        Note, the maximum number of checkpoints kept in the backend can be
+        configured at `RayConfig.num_actor_checkpoints_to_keep`.
+
+        Args:
+            actor_id: ID of the actor.
+            checkpoint_id: ID of the checkpoint that has expired.
+        Returns:
+            None.
```
Comment (pschafhalter) [inline python/ray/actor.py]: Maybe also return `True`/`False`.

File: python/ray/worker.py [hunk 23]
```diff
@@ -946,6 +972,54 @@ def _wait_for_and_process_task(self, task):
...
+        checkpoints = ray.actor.get_checkpoints_for_actor(actor_id)
+        if len(checkpoints) > 0:
+            # If we found previously saved checkpoints for this actor,
+            # call the `load_checkpoint` callback.
+            checkpoint_id = actor.load_checkpoint(actor_id, checkpoints)
+            if checkpoint_id is not None:
+                # Check that the returned checkpoint id is in the
+                # `available_checkpoints` list.
+                msg = ("`load_checkpoint must return a checkpoint id that " +
+                       "exists in the `available_checkpoints` list, or eone.")
```
Comment (pschafhalter) [inline python/ray/worker.py]: typo?

File: test/actor_test.py [hunk 24]
```diff
@@ -2380,3 +2311,259 @@ def inc(self, duration):
     for _, result_id_list in result_ids.items():
         results = list(range(1, len(result_id_list) + 1))
         assert ray.get(result_id_list) == results
+
+
+def test_checkpointing(ray_start_regular, ray_checkpointable_actor_cls):
+    """Test actor checkpointing and restoring from a checkpoint."""
+
+    def kill_actor(actor):
```
Comment (pschafhalter) [inline test/actor_test.py]: Maybe move this outside the function? The same code is repeated in a few other places.

File: python/ray/actor.py [hunk 25]
```diff
@@ -832,58 +727,142 @@ def __ray_checkpoint__(self):
...
+            checkpoint.
+
+        Returns:
+            A boolean value that indicates whether this actor needs to be
+            checkpointed.
+        """
+        pass
+
+    @abstractmethod
+    def save_checkpoint(self, actor_id, checkpoint_id):
```
Comment (pschafhalter) [inline python/ray/actor.py]: Make sure to document that we expect the user to throw an exception if saving the checkpoint fails. This only became clear to me when reading the tests.

File: python/ray/actor.py [hunk 26]
```diff
@@ -832,58 +727,142 @@ def __ray_checkpoint__(self):
...
+            checkpoint_id: ID of this checkpoint. You should save it together
+                with actor’s checkpoint data. And it will be used by the
+                `load_checkpoint` method.
+        Returns:
+            None.
+        """
+        pass
+
+    @abstractmethod
+    def load_checkpoint(self, actor_id, available_checkpoints):
```
Comment (pschafhalter) [inline python/ray/actor.py]: Maybe also add a note about throwing exceptions, or an "Excepts:" section to the docstring

**Review by pschafhalter — CHANGES_REQUESTED** [review]
A few minor comments, otherwise the Python section lgtm!

File: python/ray/actor.py (hunk 22 above)
Comment (raulchen) [reply python/ray/actor.py]: Do you want the return value to indicate whether the application checkpoint was deleted? I think the backend doesn't care about this.

File: test/actor_test.py (hunk 24 above)
Comment (raulchen) [reply test/actor_test.py]: Do you mean the similar code in the `test_actor_reconstruction` test? that one is a bit different. In that function, we need to get the pid before calling other methods. So I cannot re-use this function for that test.

**Comment by raulchen** [comment]
@pschafhalter thanks! the comments are addressed, or replied if I have questions. could you take a look again? thank you.

File: python/ray/actor.py (hunk 22 above)
Comment (pschafhalter) [reply python/ray/actor.py]: Ah, then maybe document that the user should throw an exception if the checkpoint wasn't deleted. From a user perspective, it wasn't clear to me how I should handle failing to delete a checkpoint.

File: test/actor_test.py (hunk 24 above)
Comment (pschafhalter) [reply test/actor_test.py]: It looks to me that `kill_actor` in `test_remote_checkpointing`, `test_checkpointing_save_exception`, and `test_checkpointing_load_exception` are all the same.

Comment (raulchen) [reply test/actor_test.py]: ah, I didn't notice Stephanie added those tests. Thanks.

**Comment by raulchen** [comment]
@stephanie-wang I think now we can deprecate the `__ray_checkpoint__` magic method. We can use the following way to do the same functionality:
```python
def checkpoint(self):
    self._should_checkpoint = True

def should_checkpoint(self, checkpoint_context):
    return self._should_checkpoint

actor.checkpoint.remote()
```
Also, by deprecating this, we can remove a bunch of condition checks and unneeded code.

Another small comment is that maybe we should move `_save_actor_checkpoint` and `_restore_actor_checkpoint` to `function_manager.py`. Because it's weird that they are defined as private methods of Worker class, but never used in Worker.

**Comment by stephanie-wang** [comment]
Sounds good, I can try that now (and fix the conflict too).

**Comment by raulchen** [comment]
> Sounds good, I can try that now (and fix the conflict too).

thanks! I think this PR is ready for merge if CI passes.

**Review by stephanie-wang — APPROVED** [review]
Awesome job!

**Review by pschafhalter — APPROVED** [review]
